File: 3394-check-if-grid-can-be-cut-into-sections/3394-check-if-grid-can-be-cut-into-sections.cs
public class Solution {
    public bool CheckValidCuts(int n, int[][] rectangles) {
        return CanCut(rectangles, 0) || CanCut(rectangles, 1);
    }

    private bool CanCut(int[][] rectangles, int idx){
        Array.Sort(rectangles, (r1, r2) => r1[idx] - r2[idx]);
        int cuts = 0, prevEnd = rectangles[0][idx+2];

        foreach(var rect in rectangles){
            int start = rect[idx], end = rect[idx+2];
            if(start >= prevEnd)
                if(++cuts == 2) return true;
            prevEnd = Math.Max(prevEnd, end);
        }

        return false;
    }
}

File: 3394-check-if-grid-can-be-cut-into-sections/3394-check-if-grid-can-be-cut-into-sections.py
class Solution:
    def checkValidCuts(self, n: int, rectangles: List[List[int]]) -> bool:
        return self.canCut(rectangles, 0) or self.canCut(rectangles, 1)

    def canCut(self, rectangles: List[List[int]], idx: int) -> bool:
        rectangles.sort(key=lambda r : r[idx])
        cuts = 0
        prevEnd = rectangles[0][idx+2]

        for rect in rectangles:
            start, end = rect[idx], rect[idx+2]
            if start >= prevEnd:
                cuts += 1
                if cuts == 2: return True
            prevEnd = max(prevEnd, end)

        return False